toolkit/tools/imagegen/diskutils/verity.go

```diff
@@ -0,0 +1,317 @@
+	TmpfsOverlaysDebugMount string
+}
+
+// AddRootVerityFilesToInitramfs adds files needed for a verity root to the initramfs
+// - workingFolder is a temporary folder to extract the initramfs to
+// - initramfsPath is the path to the initramfs
+func (v *VerityDevice) AddRootVerityFilesToInitramfs(workingFolder, initramfsPath string) (err error) {
+	var (
+		verityWorkingDirectory = filepath.Join(workingFolder, v.MappedName)
+	)
+	// Measure the disk and generate the hash and fec files
+	err = v.createVerityDisk(verityWorkingDirectory)
+	if err != nil {
+		return fmt.Errorf("failed while generating a verity disk: %w", err)
+	}
+
+	// Now place them in the initramfs
+	logger.Log.Info("Adding dm-verity read-only root files into initramfs")
+	initramfs, err := OpenInitramfs(initramfsPath)
+	defer initramfs.Close()
+	if err != nil {
+		return fmt.Errorf("failed to open the initramfs: %w", err)
+	}
+
+	verityFiles, err := ioutil.ReadDir(verityWorkingDirectory)
+	if err != nil {
+		return
+	}
+
+	for _, file := range verityFiles {
+		filePath := filepath.Join(verityWorkingDirectory, file.Name())
+		logger.Log.Debugf("Adding '%s' to initramfs", filePath)
+		// Place each file in the root of the initramfs
+		err = initramfs.AddFileToInitramfs(filePath, file.Name())
+		if err != nil {
+			return fmt.Errorf("failed adding %s to initramfs: %w", filePath, err)
+		}
+	}
+
+	return
+}
+
+func (v *VerityDevice) createVerityDisk(verityDirectory string) (err error) {
+	const (
+		hexChars   = "0123456789abcdef"
+		saltLength = 64
+		hashAlg    = "sha256"
+	)
+	var (
+		verityFecArgs    []string
+		verityArgs       []string
+		verityVerifyArgs []string
+		fileBase         = filepath.Join(verityDirectory, v.MappedName)
+		rootHashPath     = fmt.Sprintf("%s.roothash", fileBase)
+		hashtreePath     = fmt.Sprintf("%s.hashtree", fileBase)
+		fecFilePath      = fmt.Sprintf("%s.fec", fileBase)
+	)
+
+	err = os.MkdirAll(verityDirectory, os.ModePerm)
+	if err != nil {
+		return
+	}
+
+	rootHashFile, err := os.Create(rootHashPath)
+	if err != nil {
+		return
+	}
+	defer rootHashFile.Close()
+
+	salt, err := randomization.RandomString(64, hexChars)
+	if err != nil {
+		return
+	}
+
+	if v.FecRoots > 0 {
+		verityFecArgs = []string{
+			fmt.Sprintf("--fec-device=%s", fecFilePath),
+			fmt.Sprintf("--fec-roots=%d", v.FecRoots),
+		}
+	}
+
+	verityArgs = []string{
+		"--salt",
+		salt,
+		"--hash",
+		hashAlg,
+		"--verbose",
+		"--debug",
+		"format",
+		v.MappedDevice,
+		hashtreePath,
+	}
+
+	logger.Log.Info("Generating a dm-verity read-only partition, this may take several minutes")
+	verityOutput, stderr, err := shell.Execute("veritysetup", append(verityFecArgs, verityArgs...)...)
+	if err != nil {
+		err = fmt.Errorf("Unable to create verity disk '%s': %w", stderr, err)
+		return
+	}
+
+	// Searches for a line like: "Root hash:      1234567890abcdefg..."
+	rootHashLine, stderr, err := shell.ExecuteWithStdin(verityOutput, "grep", "^Root hash:")
```

Comment:
I think we should avoid calling grep for regex matching here. In general we use golang's built in regex engine. This avoids external dependencies and also provides the beneifits that come with calling into typed native code.

Reply:
Fair point, this was translated from a shell script. The regex actually simplifies this as well.

---
In reply to: [553013709](https://github.com/microsoft/CBL-Mariner/pull/483#discussion_r553013709) [](ancestors = 553013709)